presto-main/src/test/java/io/prestosql/execution/buffer/BenchmarkPagesSerde.java

```diff
@@ -0,0 +1,217 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.execution.buffer;
+
+import com.google.common.collect.ImmutableList;
+import io.prestosql.spi.Page;
+import io.prestosql.spi.PageBuilder;
+import io.prestosql.spi.block.BlockBuilder;
+import io.prestosql.spi.type.RowType;
+import io.prestosql.spi.type.Type;
+import io.prestosql.spiller.AesSpillCipher;
+import org.openjdk.jmh.annotations.Benchmark;
+import org.openjdk.jmh.annotations.BenchmarkMode;
+import org.openjdk.jmh.annotations.Fork;
+import org.openjdk.jmh.annotations.Measurement;
+import org.openjdk.jmh.annotations.Mode;
+import org.openjdk.jmh.annotations.OutputTimeUnit;
+import org.openjdk.jmh.annotations.Param;
+import org.openjdk.jmh.annotations.Scope;
+import org.openjdk.jmh.annotations.Setup;
+import org.openjdk.jmh.annotations.State;
+import org.openjdk.jmh.annotations.Warmup;
+import org.openjdk.jmh.infra.Blackhole;
+import org.openjdk.jmh.runner.Runner;
+import org.openjdk.jmh.runner.RunnerException;
+import org.openjdk.jmh.runner.options.Options;
+import org.openjdk.jmh.runner.options.OptionsBuilder;
+import org.openjdk.jmh.runner.options.VerboseMode;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Optional;
+import java.util.Random;
+
+import static io.airlift.slice.Slices.utf8Slice;
+import static io.prestosql.metadata.MetadataManager.createTestMetadataManager;
+import static io.prestosql.spi.type.BigintType.BIGINT;
+import static io.prestosql.spi.type.VarcharType.VARCHAR;
+import static java.nio.charset.StandardCharsets.ISO_8859_1;
+import static java.util.concurrent.TimeUnit.SECONDS;
+
+@State(Scope.Thread)
+@OutputTimeUnit(SECONDS)
+@Fork(1)
+@Warmup(iterations = 10, time = 1, timeUnit = SECONDS)
+@Measurement(iterations = 10, time = 1, timeUnit = SECONDS)
+@BenchmarkMode(Mode.Throughput)
+public class BenchmarkPagesSerde
+{
+    @Benchmark
+    public void serialize(BenchmarkData data, Blackhole bh)
+    {
+        Page[] pages = data.dataPages;
+        PagesSerde serde = data.serde;
+        for (Page p : pages) {
+            bh.consume(serde.serialize(p));
+        }
+    }
+
+    @Benchmark
+    public void deserialize(BenchmarkData data, Blackhole bh)
+    {
+        SerializedPage[] pages = data.serializedPages;
+        PagesSerde serde = data.serde;
+        for (SerializedPage p : pages) {
+            bh.consume(serde.deserialize(p));
+        }
+    }
+
+    @State(Scope.Thread)
+    public static class BenchmarkData
+    {
+        private static final int ROW_COUNT = 10000;
+        private static final RowType rowType = RowType.anonymous(ImmutableList.of(VARCHAR, VARCHAR, VARCHAR, VARCHAR));
```

Comment:
use different types (instead of just `VACHAR`)

Reply:
For this particular PR I don't really care about the types being used except that I can control the compressibility to exercise both the compression succeeds path vs the compression fails path when serializing the page (as written and using the current seed value, 1/2 of pages compress successfully).

As long as the serialization time is constant, then they serve their purpose for now, but I agree that we should have the ability exercise more aspects of serialization and deserialization performance. Maybe we can add the other cases to`BenchmarkPagesSerde` after this PR?

Reply:
> Maybe we can add the other cases toBenchmarkPagesSerde after this PR?

What about the snippet I attached in comment? Could we use it as a base (it should be easy to extend it with other block types)?

Reply:
I completely agree that we should be able to use `BenchmarkPagesSerde` to benchmark different block encodings and data types in the future. A quick look at your example seems ok, but not necessary for this PR. I see a couple problems:
1. Benchmarking the performance of buffer reuse is very sensitive to the size of the pages being serialized / deserialized, the compressibility of the data, and the number of pages serialized in each batch. You'll need a way to control those factors explicitly and target "realistic" scenarios in those dimensions.
2. Don't use `TestingPagesSerdeFactory().createPagesSerde()` to create the `PagesSerde` since it wraps all of the operations in `synchronized` blocks

I'd still rather land this first and then come back to enhance it with a full suite of types, channel counts, and encodings.

Reply:
The problem with the current setup is that we are testing a very specific case (4 varchar columns, not too many rows). It's hard to judge how much it helps in general case. It's also not obvious how to extend this benchmarks for different types which makes this benchmark sealed for single particular use-case.

Why there are 4 columns? Wouldn't one suffice?

Reply:
Let's make it use single `VARCHAR` column.

Reply:
Changed to a single `VARCHAR` column. Benchmarks still show the same basic shape modulo the new page size constraints. I had a much harder time picking values that would compress ~1/2 of the pages the way that the previous iteration did, so in this version almost all (26/30) pages compress successfully.